src/text/line.rs

```diff
@@ -586,31 +553,68 @@ impl WidgetRef for Line<'_> {
+        let line_width = self.width() as u16;
+        if line_width <= area.width {
+            let offset = match self.alignment {
+                Some(Alignment::Center) => (area.width.saturating_sub(line_width)) / 2,
+                Some(Alignment::Right) => area.width.saturating_sub(line_width),
+                Some(Alignment::Left) | None => 0,
+            };
+            render_spans(&self.spans, area, buf, offset, 0);
         } else {
             let offset = match self.alignment {
-                Some(Alignment::Center) => (area.width.saturating_sub(width)) / 2,
-                Some(Alignment::Right) => area.width.saturating_sub(width),
+                Some(Alignment::Center) => (line_width.saturating_sub(area.width)) / 2,
+                Some(Alignment::Right) => line_width.saturating_sub(area.width),
                 Some(Alignment::Left) | None => 0,
             };
-            x = x.saturating_add(offset);
-            self.to_owned()
+            render_spans(&self.spans, area, buf, 0, offset);
         };
-        for span in &line.spans {
-            let span_width = span.width() as u16;
-            let span_area = Rect {
-                x,
-                width: span_width.min(area.right().saturating_sub(x)),
-                ..area
-            };
-            span.render(span_area, buf);
-            x = x.saturating_add(span_width);
-            if x >= area.right() {
-                break;
-            }
+    }
+}
+
+/// Renders all the spans of the line that are after the `span_offset` in the line and that are
+/// visible in the given area (starting rendering at `area.x + buf_x_offset`). `buf_x_offset`
+/// probably should just be used to modify area.x instead of being a parameter, but it's here
+/// because that would have made the above code just a little uglier
+///
+/// Algorithm:
+/// 1. For each span in the line:
+///    1. If the span is completely before the `span_offset`, skip it
+///    2. If the span is partially before the `span_offset`, render the right side of the span
+///    3. If the span is completely after the `span_offset`, render the whole span
+/// 2. Update `buf_x_offset` to account for the width of the rendered span
+/// 3. Update `span_offset` to 0 after the first span is rendered in full
+/// 4. Repeat
+fn render_spans(
+    spans: &[Span],
+    area: Rect,
+    buf: &mut Buffer,
+    mut buf_x_offset: u16,
+    mut span_offset: u16,
+) {
+    for span in spans {
+        let span_width = span.width() as u16;
+        // ignore spans that are completely before the offset
+        if span_offset >= span_width {
+            span_offset -= span_width;
+            continue;
+        }
+        let area = Rect {
+            x: area.x.saturating_add(buf_x_offset),
+            width: area.width.saturating_sub(buf_x_offset),
+            ..area
+        };
+        let is_span_partially_visible = span_offset > 0;
+        if is_span_partially_visible {
```

Comment:
When first reading this I thought "oh, the span is visible, so it should be >=?"
Inverting the `if` might clarify this: `fully_visible = span_offset == 0` → else only partial.

Reply:
I prefer this direction for clarity as it any partial rendered span comes before all the fully rendered ones. Top to bottom reading order helps convey the order of spans.

Reply:
Then improve the variable name to make this more clear
```suggestion
        let span_is_only_partially_visible = span_offset > 0;
        if span_is_only_partially_visible {
```

Reply:
Actually, I'll change this on second thoughts.

Reply:
Actually - I don't like the clippy lint that occurs here (or the allow that it necessitates) - it requires flipping the if / else blocks.

Reply:
Suggestion applied.